File: examples/canonical-analytics/go/internal/charts/time_series.go
package charts

import (
	"time"

	"analytics-workbench-go/internal/analytics"
)

func BuildTimeSeriesFigure(agg analytics.EventCountSeries) PlotlyFigure {
	now := time.Now().UTC().Format(time.RFC3339)

	if len(agg.Dates) == 0 {
		return PlotlyFigure{
			Data: []PlotlyTrace{{
				Type:          "scatter",
				Mode:          "lines+markers",
				Name:          "empty",
				HoverTemplate: "No data",
				X:             []string{},
				Y:             []int{},
			}},
			Layout: PlotlyLayout{
				Title: PlotlyTitle{Text: "No data for selected filters"},
			},
			Meta: PlotlyMeta{
				TotalCount:  0,
				GeneratedAt: now,
			},
		}
	}

	var traces []PlotlyTrace
	for env, counts := range agg.ByEnvironment {
		traces = append(traces, PlotlyTrace{
			Type:          "scatter",
			Mode:          "lines+markers",
			Name:          env,
			X:             agg.Dates,
			Y:             counts,
			HoverTemplate: "%{x}: %{y} events<extra></extra>",
		})
	}

	var total int
	for _, c := range agg.Counts {
		total += c
	}

	return PlotlyFigure{
		Data: traces,
		Layout: PlotlyLayout{
			Title: PlotlyTitle{Text: "Event Volume Trends"},
			XAxis: PlotlyAxis{Title: PlotlyAxisTitle{Text: "Date"}},
			YAxis: PlotlyAxis{Title: PlotlyAxisTitle{Text: "Event Count"}},
		},
		Meta: PlotlyMeta{
			TotalCount:  total,
			GeneratedAt: now,
		},
	}
}

File: examples/canonical-analytics/elixir/analytics_workbench/lib/analytics_workbench/charts/time_series.ex
defmodule AnalyticsWorkbench.Charts.TimeSeries do
  def build_figure(agg) do
    data = Enum.map(agg.by_environment, fn {env, counts} ->
      %{
        type: "scatter",
        mode: "lines+markers",
        name: env,
        x: agg.dates,
        y: counts,
        hovertemplate: "%{y} events<br>%{x}<extra></extra>"
      }
    end)

    total_count = Enum.sum(agg.counts)

    %{
      data: data,
      layout: %{
        title: %{text: "Event Volume Over Time"},
        xaxis: %{title: %{text: "Date"}},
        yaxis: %{title: %{text: "Event Count"}}
      },
      meta: %{
        total_count: total_count,
        generated_at: DateTime.utc_now() |> DateTime.to_iso8601()
      }
    }
  end
end
